feat: Oncall integration labels

willgallego-grafana opened the pull request:
Creating a label datasource, we can pass a label key/value pair to the `labels` attribute for a `grafana_oncall_integration` to enable labels through terraform.

Sample terraform:

```terraform
terraform {
  required_providers {
    grafana = { 
      source  = "grafana/grafana"
      version = ">= 1.22.0"
    }   
  }
}

provider "grafana" {
  alias      = "oncall"
  url        = "http://grafana:3000"
  oncall_url = "http://oncall-engine:8080"
  auth       = "<token here>"
}

// Step 2: Interact with Grafana OnCall
resource "grafana_oncall_integration" "test-acc-integration" {
  provider = grafana.oncall
  name     = "my integration"
  type     = "webhook"
  default_route {}
  labels = [data.grafana_oncall_label.test-label]
}

data "grafana_oncall_label" "test-label" {
  provider = grafana.oncall
  key      = "KeyInTerraform"
  value    = "ValueInTerraform"
}

output "example" {
  value = data.grafana_oncall_label.test-label
}

output "key_value_list" {
  value = [data.grafana_oncall_label.test-label]
}

output "integration" {
  value = resource.grafana_oncall_integration.test-acc-integration
}
```

willgallego-grafana committed:
adding data source grafana_oncall_label

willgallego-grafana committed:
WiP

labels schema isn't quite there yet, but getting there!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/grafana/terraform-provider-grafana?pullRequest=2125) <br/>All committers have signed the CLA.

willgallego-grafana committed:
WiP - integration labels

This feels closer to what we want - a list of objects. It's not quite
being parsed correctly, though, and I had to comment out its use in
resourceIntegrationCreate

willgallego-grafana committed:
Using TypeMap in resource integration label "labels"

willgallego-grafana commented on internal/resources/oncall/data_source_label.go:
```diff
@@ -0,0 +1,66 @@
+	var data labelDataSourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+
+	data.ID = basetypes.NewStringValue(data.Key.ValueString())
```
TODO - I probably need a better way of setting the id here. Our modeling in the app has the uniqueness derived from the key/value pairing, not just the key itself (though I don't know if we care about that in the terraform code)

willgallego-grafana committed:
cleaning up with go fmt, fixing updates

Updates work now. Alongside that, ran a go fmt to take care of some
clean up and removed some commented code.

willgallego-grafana committed:
WiP: tweaking the flattened_labels

It's not quite right yet, need to test it better, but progress

willgallego-grafana committed:
putting in a check for key/value for labels

Each label should have both a "key" and a "value" assigned. We can't
force Terraform's schema to do that using TypeMap.

What we can do, though, is check they exist and skip the label then.
It'd be better to warn the user or error on it, but this is better than
a panic.

willgallego-grafana commented on internal/resources/oncall/resource_integration.go:
```diff
@@ -301,13 +318,15 @@ func resourceIntegrationCreate(ctx context.Context, d *schema.ResourceData, clie
 		Type:         typeData,
 		Templates:    expandTemplates(templatesData),
 		DefaultRoute: expandDefaultRoute(defaultRouteData),
+		Labels:       expandLabels(labelsData),
```
Until the `amixr-api-go-client` changes are pushed, these will continue to fail.

willgallego-grafana commented on internal/resources/oncall/resource_integration.go:
```diff
@@ -792,3 +814,36 @@ func expandDefaultRoute(input []interface{}) *onCallAPI.DefaultRoute {
+
+	for _, r := range input {
+		inputMap := r.(map[string]interface{})
+		key, keyExists := inputMap["key"]
```
This is a work around, of sorts.

We're using TypeMap here and the preferred way for users to interact with labels as part of oncall integrations is to create a data source object which enforces requiring `key` and `value` as attributes, then passing that object into the list.

A user could, instead, create the object directly within the list such as:
```
...
labels = [
    {
      key = "SomeKey"
      value = "SomeValue"
    }
]
...
```

which would work. That said, `schema.TypeMap` can't enforce required values (see: https://developer.hashicorp.com/terraform/plugin/sdkv2/schemas/schema-types#typemap) and if the user leaves one out, it would normally panic.

This is at least a check so it doesn't totally fall over, but it isn't great UX as it doesn't indicate to the user they've left anything out. We could rework the returns on the `flattenLabels` and other checks such that we can return an error in the `diag` object.

willgallego-grafana committed:
fixing flattened_labels

willgallego-grafana committed:
adding test for data_source_label

willgallego-grafana committed:
Merge branch 'main' into oncallIntegrationLabels

willgallego-grafana committed:
adding tests for grafana oncall integration w/ labels

willgallego-grafana committed:
Merge branch 'main' into oncallIntegrationLabels

willgallego-grafana committed:
formatting

willgallego-grafana committed:
updating docs

willgallego-grafana committed:
fixing datasource label test

willgallego-grafana committed:
changing how id is populated

a label is unique by the key/value pair, so the id for that should
reflect that fact.

We could potentially do a hash on this as well, but this is probably ok
for now. As colons aren't legal characters, this should remain unique
without issues like the key or value including it.

willgallego-grafana committed:
fixing missing fmt

willgallego-grafana committed:
clean up data source after test

willgallego-grafana committed:
reverting id change

having issues with CI/CD, trying to clear
